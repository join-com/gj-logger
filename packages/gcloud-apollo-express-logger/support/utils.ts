export const pick = (obj: any, keys: any[]) => {
  return keys
    .map(k => (k in obj ? { [k]: obj[k] } : {}))
    .reduce((acc, cur) => Object.assign(acc, cur), {});
};
