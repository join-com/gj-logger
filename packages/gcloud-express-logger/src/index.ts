import gcloudLogger, { Logger } from '@join-com/gcloud-logger';
import { NextFunction, Request, Response } from 'express';
import onFinished from 'on-finished';

const EXCLUDED_KEYS = ['file', 'resume'];
const requestLogMessage = (req: Request, res: Response, ms: number) => ({
  httpRequest: {
    remoteIp: req.ip,
    requestUrl: req.originalUrl || req.url,
    requestMethod: req.method,
    status: res.statusCode,
    userAgent: req.headers['user-agent'],
    referer: req.headers.referer || req.headers.referrer,
    latency: `${ms.toFixed(3)}ms`,
    path: req.route && req.route.path,
  },
  query: req.query,
  reqBody: filterBody(req.body),
  requestTime: ms,
});

const filterBody = (body: any) => {
  Object.keys(body).forEach(key => {
    if (EXCLUDED_KEYS.indexOf(key) !== -1) {
      return '[EXCLUDED]';
    }

    if (typeof body[key] === 'object') {
      filterBody(body[key]);
    }
  });
  return body;
};

export const requestLogger = (
  logger: Logger = gcloudLogger,
  startTrace?: (req: Request) => void,
) => (req: Request, res: Response, next: NextFunction) => {
  const startTime = process.hrtime();

  const logRequest = () => {
    if (startTrace) {
      startTrace(req);
    }
    const diff = process.hrtime(startTime);
    const ms = diff[0] * 1e3 + diff[1] * 1e-6;
    const message = requestLogMessage(req, res, ms);
    if (res.statusCode >= 500) {
      logger.error(req.path, message);
    } else if (res.statusCode >= 400) {
      logger.warn(req.path, message);
    } else {
      logger.info(req.path, message);
    }
  };
  if (req.originalUrl !== '/healthz' && req.originalUrl !== '/readiness') {
    onFinished(res, logRequest);
  }
  next();
};

export { NextFunction, Request, Response } from 'express';
export default requestLogger;
