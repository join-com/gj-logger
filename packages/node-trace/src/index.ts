import * as asyncHooks from 'async_hooks';
import * as uuid from 'uuid';

export const DEFAULT_TRACE_CONTEXT_NAME = 'x-cloud-trace-context';

global.__join_com_node_trace__ = new Map();

const init = (asyncId: number, _: string, triggerAsyncId: number) => {
  if (!global.__join_com_node_trace__.has(triggerAsyncId)) {
    return;
  }
  global.__join_com_node_trace__.set(
    asyncId,
    global.__join_com_node_trace__.get(triggerAsyncId),
  );
};

const destroy = (asyncId: number) => {
  global.__join_com_node_trace__.delete(asyncId);
};

asyncHooks.createHook({ init, destroy }).enable();

const generateTrace = (): string => `${uuid.v4()}/0`;

export const start = (traceId: string = generateTrace()) => {
  global.__join_com_node_trace__.set(asyncHooks.executionAsyncId(), traceId);
};

export const getTraceContext = (): string => {
  return global.__join_com_node_trace__.get(asyncHooks.executionAsyncId());
};

export const getTraceContextName = (): string => {
  return process.env.TRACE_CONTEXT_NAME || DEFAULT_TRACE_CONTEXT_NAME;
};

export { traceMiddleware } from './expressMiddleware';

export default {
  DEFAULT_TRACE_CONTEXT_NAME,
  start,
  getTraceContext,
  getTraceContextName,
};
