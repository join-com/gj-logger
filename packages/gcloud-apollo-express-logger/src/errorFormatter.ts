import { GraphQLError } from 'graphql/error/GraphQLError';
import { ClassValidationError } from '@join-private/base-errors';
import { pick, omit } from '../support/utils';

interface IFormatterOptions {
  whiteList?: string[];
  blackList?: string[];
}

interface IException {
  code?: number;
}

export const errorFormatter = (options?: IFormatterOptions) => (
  error: GraphQLError,
) => {
  if (error.extensions) {
    let exception = error.extensions.exception;
    if (exception && exception.validationErrors) {
      exception = new ClassValidationError(exception.validationErrors);
    } else if (isAuthenticationError(error)) {
      exception = {
        code: 'unauthorized',
        message: 'Authentication error',
      };
    } else if (isForbiddenError(error)) {
      exception = {
        code: 'forbidden',
        message: 'Forbidden error',
      };
    } else if (isUnknownError(exception)) {
      exception = {
        code: 'unknown',
        message: 'Server error',
      };
    }
    if (options) {
      const { whiteList, blackList } = options;
      exception = whiteList ? pick(exception, whiteList) : exception;
      exception = blackList ? omit(exception, blackList) : exception;
    }
    error.extensions.exception = exception;
  }

  return error;
};

const isUnknownError = (exception: IException | undefined) =>
  Boolean(!exception || (exception && !exception.code));

const isAuthenticationError = (error: GraphQLError) =>
  Boolean(error.extensions && error.extensions.code === 'UNAUTHENTICATED');

const isForbiddenError = (error: GraphQLError) =>
  Boolean(error.extensions && error.extensions.code === 'FORBIDDEN');
