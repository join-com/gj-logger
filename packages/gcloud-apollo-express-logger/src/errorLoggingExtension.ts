import logger, { Level } from '@join-com/gcloud-logger-trace';
import * as trace from '@join-com/node-trace';
import { GraphQLExtension } from 'apollo-server-core';
import { GraphQLError } from 'graphql';
import { ForbiddenError, Maybe } from 'type-graphql';

export default errorLoggingExtension;

const isAboveWarningLevel = (originalError: Maybe<Error>): boolean =>
  !(originalError instanceof ForbiddenError);

const determineLogLevel = (errors: readonly GraphQLError[]): Level => {
  const errorsAboveWarningLevel = errors
    .map(e => e.originalError)
    .filter(isAboveWarningLevel);
  return errorsAboveWarningLevel.length ? Level.ERROR : Level.WARNING;
};

export const errorLoggingExtension: GraphQLExtension = {
  requestDidStart: ({ request }) => {
    if (
      !request ||
      !request.headers ||
      !request.headers.get(trace.getTraceContextName())
    ) {
      logger.warn("No trace id present - can't enable request tracing");
      return;
    }
    trace.start(request.headers.get(trace.getTraceContextName())!);
  },
  didEncounterErrors: errors => {
    const errorMessages = errors.map(e => e.message).join(', ');
    logger.log(
      determineLogLevel(errors),
      `Encountered errors when processing GraphQL request: [${errorMessages}]`,
      errors,
    );
  },
};
