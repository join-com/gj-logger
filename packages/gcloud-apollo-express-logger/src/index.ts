import logger, { Level } from '@join-com/gcloud-logger-trace';
import * as trace from '@join-com/node-trace';
import { ClassValidationError } from '@join-private/base-errors';
import { GraphQLExtension } from 'apollo-server-core';
import { GraphQLError } from 'graphql';
import { ForbiddenError, Maybe } from 'type-graphql';
import { pick, omit } from '../support/utils';

interface IFormatterOptions {
  whiteList?: string[];
  blackList?: string[];
}

interface IException {
  code?: number;
}

export const errorFormatter = (options?: IFormatterOptions) => (
  error: GraphQLError,
) => {
  if (error.extensions) {
    let exception = error.extensions.exception;
    if (exception?.validationErrors) {
      exception = new ClassValidationError(exception.validationErrors);
    } else if (isAuthorizationError(error)) {
      exception = {
        code: 401,
        message: 'Authorization error',
      };
    } else if (isForbiddenError(error)) {
      exception = {
        code: 403,
        message: 'Forbidden error',
      };
    } else if (isUnknownError(exception)) {
      exception = {
        code: 500,
        message: 'Server error',
      };
    }
    if (options) {
      const { whiteList, blackList } = options;
      exception = whiteList ? pick(exception, whiteList) : exception;
      exception = blackList ? omit(exception, blackList) : exception;
    }
    error.extensions.exception = exception;
  }

  return error;
};

const isUnknownError = (exception: IException | undefined) =>
  Boolean(!exception || (exception && !exception.code));

const isAuthorizationError = (error: GraphQLError) =>
  Boolean(error.name === 'UnauthorizedError');

const isForbiddenError = (error: GraphQLError) =>
  Boolean(error.name === 'ForbiddenError');

const isAboveWarningLevel = (originalError: Maybe<Error>): boolean =>
  !(originalError instanceof ForbiddenError);

const determineLogLevel = (errors: readonly GraphQLError[]): Level => {
  const errorsAboveWarningLevel = errors
    .map(e => e.originalError)
    .filter(isAboveWarningLevel);
  return errorsAboveWarningLevel.length ? Level.ERROR : Level.WARNING;
};

export const errorLoggingExtension: GraphQLExtension = {
  requestDidStart: ({ request }) => {
    if (
      !request ||
      !request.headers ||
      !request.headers.get(trace.getTraceContextName())
    ) {
      logger.warn("No trace id present - can't enable request tracing");
      return;
    }
    trace.start(request.headers.get(trace.getTraceContextName())!);
  },
  didEncounterErrors: errors => {
    const errorMessages = errors.map(e => e.message).join(', ');
    logger.log(
      determineLogLevel(errors),
      `Encountered errors when processing GraphQL request: [${errorMessages}]`,
      errors,
    );
  },
};
export default errorLoggingExtension;
