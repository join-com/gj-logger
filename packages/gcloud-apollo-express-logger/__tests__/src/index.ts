import { formatError, errorLoggingExtension } from '../../src/index';
import { ApolloServer } from 'apollo-server-express';
import express from 'express';
import request from 'supertest';
import { GraphQLError } from 'graphql';
import { buildSchema } from 'type-graphql';
import { TestResolver } from '../../support/TestResolver';
import { DEFAULT_TRACE_CONTEXT_NAME, start } from '@join-com/node-trace';

jest.mock('@join-com/node-trace', () => ({
  ...jest.requireActual('@join-com/node-trace'),
  start: jest.fn(),
}));

describe('formatError', () => {
  it('formats error object', () => {
    const graphqlError = new GraphQLError('message');
    const formattedError = formatError(graphqlError);
    expect(formattedError).toMatchSnapshot();
  });
});

describe('errorLoggingExtension', () => {
  let server, app: express.Express;
  beforeAll(async () => {
    const schema = await buildSchema({
      resolvers: [TestResolver],
    });
    server = new ApolloServer({
      schema,
      extensions: [() => errorLoggingExtension],
    });
    app = await express();
    server.applyMiddleware({ app, cors: true, path: '/test' });
  });
  it('Should report errors', async () => {
    const query = `
        query Test {
          testValidationError {
              id
          }
        }
      `;
    await request(app)
      .post('/test')
      .set(DEFAULT_TRACE_CONTEXT_NAME, 'someTraceId')
      .send({ query })
      .expect(400);
    expect(start).toHaveBeenCalledWith('someTraceId');
  });
});
