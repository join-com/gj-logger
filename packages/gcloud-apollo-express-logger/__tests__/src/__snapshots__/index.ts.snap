// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatError formats error object 1`] = `[GraphQLError: message]`;
